### ISSUE TITLE
IWYU suggest to use the header where forward-declare is enough

### ISSUE THREAD

**Issue by gdebast** [include-what-you-use/include-what-you-use#1501]
Hello,

I'm currently investigating the usage of IWYU for our company which develop a C++ based software. I built IWYU.exe with LLVM version 16 and IWYU branch "clang_16". I'm testing it manually on our code base and I have detected what looks like a minor bug. It is a minor bug because what it suggests is just sub-optimal for the compiler. Indeed, it suggests to replace a forward-declare by an include, which can trigger additional unnecessary compilation.

This is issue is very similar to issue number [1499](https://github.com/include-what-you-use/include-what-you-use/issues/1499)

Please tell me if you need other information
Thanks already for reading
Gatien

**How to reproduce:**

Command:
`include-what-you-use.exe iwyu_objecttracker.cpp`

iwyu_objecttracker.cpp:
```
#include "iwyu_objecttracker.h"

#include "iwyu_object.h"

using namespace IWYU;

ObjectTracker::ObjectTracker() = default;
ObjectTracker::~ObjectTracker() = default;

auto ObjectTracker::make() -> Object *
{
   m_trackingObjects[0][0] = std::make_unique<Object>();
   return m_trackingObjects[0][0].get();
}
```

iwyu_objecttracker.h:
```
#pragma once

#include <memory>
#include <unordered_map>

namespace IWYU
{
   class Object;
}

namespace IWYU
{
   class ObjectTracker
   {
   public:
      ObjectTracker();
      ~ObjectTracker();

      [[nodiscard]] auto make() -> Object *;

   private:
      std::unordered_map<int, std::unordered_map<int, std::unique_ptr<Object>>> m_trackingObjects;
   };
}
```

iwyu_object.h:
```
#pragma once

namespace IWYU
{
   class Object
   {
   public:
      Object() = default;
      ~Object() = default;
   };
}
```

report:
```
iwyu_objecttracker.h should add these lines:
#include "iwyu_object.h"  // for Object

iwyu_objecttracker.h should remove these lines:
- namespace IWYU { class Object; }  // lines 8-8

The full include-list for iwyu_objecttracker.h:
#include <memory>         // for unique_ptr
#include <unordered_map>  // for unordered_map
#include "iwyu_object.h"  // for Object
---

(iwyu_objecttracker.cpp has correct #includes/fwd-decls)
```

Issue:
To my knowledge, replacing the forward-declare `namespace IWYU { class Object; }` an include `#include "iwyu_object.h"` is unnecessary and sub-optimal for the compiler. Indeed, with that suggestion, if `iwyu_objecttracker.h` is included in another cpp called `iwyu_other.cpp`, and if `iwyu_object.h` is changed, then the compilation step is triggered for `iwyu_other.cpp`, while it would have not if IWYU::Object was only forward-declared.
In that sense, this suggestion is sub-optimal for the compilation stage.

**Comment by kimgr** [comment]
@gdebast Yeah, I agree a forward decl would be preferable in this case. Conceptually, the size of `unordered_map<int, unordered_map<int, std::unique_ptr<X>>>` should not be affected by the type of X, and both construction and destruction is out-of-line.

Upping the verbosity level to 6 points to this function: https://github.com/include-what-you-use/include-what-you-use/blob/e7aa3cbc46b0edd9a1616c8d7ffde2a09782d067/iwyu.cc#L3164.

I honestly can't get my head around that -- not sure where the decision is made for fwd decl vs. full use, but I see we're landing in the very last leg of `ReportTypeUseInternal`: https://github.com/include-what-you-use/include-what-you-use/blob/e7aa3cbc46b0edd9a1616c8d7ffde2a09782d067/iwyu.cc#L3164

**Comment by kimgr** [comment]
Should have been fixed on mainline by #1588

**Comment by gdebast** [comment]
Thank you :)
